File: whatsapp-mcp-server/tool_policy.py
"""Operation-level access control (read-only mode, per-tool allow/deny).

``WHATSAPP_ALLOWED_CHATS`` (chat_policy.py) restricts *which chats* an agent may
touch. This module restricts *what it may do*. ``WHATSAPP_READ_ONLY=1`` turns the
server into a read-and-draft deployment: every tool with an external side effect
is removed from ``tools/list`` — so the model never sees it — and refuses with the
standard ``denied`` envelope if it is called anyway.

Two more knobs cut finer, by tool name:

- ``WHATSAPP_DENY_TOOLS=delete_message,leave_group`` — never offer these.
- ``WHATSAPP_ALLOW_TOOLS=list_messages,search_contacts,send_reaction`` — offer
  *only* these (an allow-list over every tool, reads included).

The three filters only ever remove capability: deny wins over allow, and
read-only wins over both, so ``WHATSAPP_ALLOW_TOOLS=send_message`` cannot switch
sending back on while read-only is set. Unknown tool names are a startup error
listing the valid ones — a typo in an allow-list must not silently widen it.

That is the right default whenever the agent reads attacker-controlled text
(any group, any forwarded message): a prompt injection can then ask for a send,
but there is nothing to call. The bridge enforces all three variables on its
mutating REST endpoints (403, whatsapp-bridge/read_only.go and tool_policy.go,
which maps each tool name to the endpoint it calls) as a second line of defence,
so a bug or a bypass on this side still cannot reach WhatsApp.

Which tools count as mutating is not a list maintained by hand: the
``@mutating_tool`` decorator in main.py registers each one, so a new tool is
either decorated (and covered) or it is not (and stays readable).

Deliberately *not* mutating:

- ``download_media`` / ``read_media`` / ``transcribe_audio`` — fetch and read;
  they only write to the local media cache. Naming ``download_media`` in a list
  still binds the other two: both fetch an uncached file through the endpoint
  ``download_media`` calls, so ``offers_download()`` gates that implicit fetch
  and the tools fall back to what the store already holds (issue #350).
- ``annotate_media`` and ``annotate`` — write notes.db, which is local MCP-owned
  state, never WhatsApp. A read-only assistant still needs somewhere to keep its
  own notes; a triage pass that cannot record what it concluded is the failure
  issue #286 describes.

Deliberately mutating even though it reads:

- ``get_group_invite_link`` — ``reset=True`` revokes the current link, and a link
  is group access that can be leaked. Blocked wholesale rather than per-argument.
"""

from __future__ import annotations

import functools
import os
from collections.abc import Callable, Collection, Mapping, Sequence
from dataclasses import dataclass
from typing import Any, Protocol

from errors import ToolError

READ_ONLY_ENV = "WHATSAPP_READ_ONLY"
ALLOW_TOOLS_ENV = "WHATSAPP_ALLOW_TOOLS"
DENY_TOOLS_ENV = "WHATSAPP_DENY_TOOLS"

_TRUE = ("1", "true", "yes", "on")
_FALSE = ("0", "false", "no", "off")

# The tool every implicit fetch is: read_media, transcribe_audio and the ingest
# worker all reach the bridge's /api/download, the endpoint download_media
# calls. Named once here so the three ask the same question.
DOWNLOAD_TOOL = "download_media"

# Populated by @mutating_tool at import time; see module docstring.
_MUTATING: set[str] = set()


def parse_bool_env(raw: str | None, var: str) -> bool:
    """Strict boolean parse: unset is False, anything unrecognised is an error.

    A security switch must not silently fall back to "off" because it was spelled
    ``WHATSAPP_READ_ONLY=treu``.
    """
    value = (raw or "").strip().lower()
    if not value:
        return False
    if value in _TRUE:
        return True
    if value in _FALSE:
        return False
    raise ValueError(f"{var}={raw!r} is not a boolean; use one of {', '.join(_TRUE + _FALSE)}")


def parse_tool_list(raw: str | None) -> frozenset[str]:
    """Comma-separated tool names; empty entries ignored. Validation is separate."""
    return frozenset(item.strip() for item in (raw or "").split(",") if item.strip())


@dataclass(frozen=True)
class ToolPolicy:
    """Which tools this process offers.

    Each field only removes capability: ``deny`` wins over ``allow``, and
    ``read_only`` wins over both.
    """

    read_only: bool = False
    allow: frozenset[str] = frozenset()
    deny: frozenset[str] = frozenset()

    def allows(self, name: str) -> bool:
        if name in self.deny:
            return False
        if self.allow and name not in self.allow:
            return False
        return not (self.read_only and name in _MUTATING)

    def blocked(self) -> tuple[str, ...]:
        """Mutating tools blocked by read-only (allow/deny also reach read tools)."""
        return tuple(sorted(name for name in _MUTATING if not self.allows(name)))

    def denial_message(self, name: str) -> str:
        if name in self.deny:
            return f"{name} is disabled: it is listed in {DENY_TOOLS_ENV}"
        if self.allow and name not in self.allow:
            return f"{name} is disabled: {ALLOW_TOOLS_ENV} is set and does not list it"
        return (
            f"{name} is disabled: {READ_ONLY_ENV} is set, so this server may read WhatsApp "
            f"but not act on it. Report the draft to the user and let them send it."
        )

    def validate(self, known: Collection[str]) -> None:
        """Fail closed on a name that is not a tool: a typo must not widen a list.

        ``known`` is what the server has registered (``registered_tool_names``).
        """
        for var, names in ((ALLOW_TOOLS_ENV, self.allow), (DENY_TOOLS_ENV, self.deny)):
            unknown = sorted(names - set(known))
            if unknown:
                raise ValueError(
                    f"{var} lists unknown tool(s): {', '.join(unknown)}. Valid names: {', '.join(sorted(known))}"
                )

    def summary(self, removed: Sequence[str] | None = None) -> str:
        """One line for the startup log; ``removed`` comes from apply_tool_policy."""
        parts: list[str] = []
        parts.append(f"{READ_ONLY_ENV}=1: read-only" if self.read_only else f"{READ_ONLY_ENV} unset")
        if self.allow:
            parts.append(f"{ALLOW_TOOLS_ENV}: {', '.join(sorted(self.allow))}")
        if self.deny:
            parts.append(f"{DENY_TOOLS_ENV}: {', '.join(sorted(self.deny))}")
        state = "; ".join(parts)
        if removed is None:
            removed = self.blocked()
        if not removed:
            return f"Tool policy — {state}; every tool enabled"
        return f"Tool policy — {state}; {len(removed)} tool(s) hidden ({', '.join(removed)})"


def load_tool_policy(env: Mapping[str, str] | None = None) -> ToolPolicy:
    """Build the policy from the environment. Raises ValueError on a bad value."""
    source: Mapping[str, str] = os.environ if env is None else env
    return ToolPolicy(
        read_only=parse_bool_env(source.get(READ_ONLY_ENV), READ_ONLY_ENV),
        allow=parse_tool_list(source.get(ALLOW_TOOLS_ENV)),
        deny=parse_tool_list(source.get(DENY_TOOLS_ENV)),
    )


_active: ToolPolicy | None = None


def set_active_policy(policy: ToolPolicy | None) -> None:
    """Install the policy the call-time guard checks (None = resolve from env again)."""
    global _active
    _active = policy


def active_policy() -> ToolPolicy:
    """The policy in force, resolved from the environment on first use.

    ``__main__`` installs the parsed policy at startup and exits on a bad value;
    this fallback matters for an in-process import (tests, embedding), where a
    value we cannot parse means read-only — fail closed.
    """
    global _active
    if _active is None:
        try:
            _active = load_tool_policy()
        except ValueError:
            _active = ToolPolicy(read_only=True)
    return _active


def offers_download() -> bool:
    """Whether the active policy still offers ``download_media``.

    The question ``read_media``, ``transcribe_audio`` and the ingest worker ask
    before fetching a file the store does not have: that fetch is
    ``download_media`` under another name, so a deployment that took the tool
    away must not get it back through a tool it kept (issue #350).
    """
    return active_policy().allows(DOWNLOAD_TOOL)


def download_denied(caller: str) -> ToolError:
    """The ``denied`` error for an implicit fetch the policy does not allow.

    Returned rather than raised, so the caller decides where the refusal sits.
    It names ``download_media`` and the list that removed it: ``caller`` is
    still enabled, so an agent reading only "denied" would take the refusal for
    a bug in the tool it actually called.
    """
    return ToolError(
        "denied",
        f"{caller} would have to fetch this file from WhatsApp first, and "
        f"{active_policy().denial_message(DOWNLOAD_TOOL)}. Media already in the store is still readable.",
    )


def mutating_tool(fn: Callable[..., Any]) -> Callable[..., Any]:
    """Mark an MCP tool as having an external side effect.

    Registers the name as mutating and refuses the call when the active policy
    blocks it. Goes *below* ``@tool_errors`` so the refusal is returned as the
    standard ``{"error": {"code": "denied", ...}}`` envelope::

        @mcp.tool()
        @tool_errors
        @mutating_tool
        def send_message(...): ...
    """
    _MUTATING.add(fn.__name__)

    @functools.wraps(fn)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        policy = active_policy()
        if not policy.allows(fn.__name__):
            raise ToolError("denied", policy.denial_message(fn.__name__))
        return fn(*args, **kwargs)

    return wrapper


def mutating_tools() -> frozenset[str]:
    """Every tool marked with @mutating_tool (empty until main.py is imported)."""
    return frozenset(_MUTATING)


class _ToolServer(Protocol):
    """The slice of MCPServer this module uses (tests pass a fake)."""

    def remove_tool(self, name: str) -> None: ...


def registered_tool_names(server: Any) -> frozenset[str]:
    """Names currently offered by the server (SDK v2 keeps no public accessor)."""
    return frozenset(tool.name for tool in server._tool_manager.list_tools())


def apply_tool_policy(server: _ToolServer, policy: ToolPolicy) -> list[str]:
    """Unregister every blocked tool so it never reaches ``tools/list``.

    Returns the names removed, for the startup log.
    """
    registered = registered_tool_names(server)
    removed = sorted(name for name in registered if not policy.allows(name))
    for name in removed:
        server.remove_tool(name)
    return removed

File: whatsapp-bridge/tool_policy.go
package main

// Per-tool allow/deny, bridge side.
//
// WHATSAPP_ALLOW_TOOLS / WHATSAPP_DENY_TOOLS name MCP *tool* names, not
// endpoints, so one value can be written once and handed to both containers
// meaning the same thing (compose passes both variables to both services). The
// MCP server hides the blocked tools from tools/list (tool_policy.py); the
// bridge repeats the decision on the REST endpoints those tools call, so a
// direct REST caller, a bypass or a bug on the MCP side still cannot reach
// WhatsApp — the same belt-and-braces split WHATSAPP_READ_ONLY uses.
//
// The translation is endpointTools below: endpoint -> the tools that reach it.
// It is endpoint-granular, and three tools share /api/send, so an endpoint
// stays open while *any* tool that uses it is allowed; the MCP server is the
// finer filter. Tool names the bridge has nothing to enforce for (reads served
// from messages.db, the read endpoints that stay open in read-only mode, notes
// written locally) are still valid entries — unenforcedTools — because the
// list is written once for both sides.
//
// Order: WHATSAPP_READ_ONLY is evaluated first (read_only.go) and wins, then
// deny, then allow. Every filter only removes capability, so an allow-list can
// never re-open a read-only bridge. Unknown names stop the bridge at startup
// with the list of valid ones, exactly as they stop the MCP server: a typo in
// an allow-list must not silently widen it.

import (
	"fmt"
	"net/http"
	"os"
	"sort"
	"strings"
)

const (
	allowToolsEnv = "WHATSAPP_ALLOW_TOOLS"
	denyToolsEnv  = "WHATSAPP_DENY_TOOLS"
)

// endpointTools maps every mutating REST endpoint to the MCP tools that call
// it (whatsapp-mcp-server/whatsapp.py). Keys must stay in step with the
// endpoints registered through `mutate` in newRESTMux — TestEndpointToolsCoversEveryMutatingEndpoint
// fails otherwise — and the tool names with the MCP server's, which
// tests/test_bridge_tool_policy.py checks from the Python side.
var endpointTools = map[string][]string{
	"/api/send":               {"send_message", "send_file", "send_audio_message"},
	"/api/react":              {"send_reaction"},
	"/api/typing":             {"send_typing"},
	"/api/mark-read":          {"mark_messages_read"},
	"/api/delete":             {"delete_message"},
	"/api/edit":               {"edit_message"},
	"/api/forward":            {"forward_message"},
	"/api/group/participants": {"manage_group_participants"},
	"/api/group/subject":      {"update_group"},
	"/api/group/invite":       {"get_group_invite_link"},
	"/api/group/leave":        {"leave_group"},
	"/api/media/purge":        {"purge_media"},
	"/api/history":            {"request_history"},
}

// unenforcedTools are the remaining MCP tool names: reads (including the ones
// served by /api/group/members, /api/poll and /api/download, which stay open
// here just as they do in read-only mode) and writes that never leave the host.
// Accepted in both lists, enforced by the MCP server only.
var unenforcedTools = []string{
	"annotate",
	"annotate_media",
	"bridge_status",
	"compact",
	"coverage",
	"download_media",
	"export_messages",
	"get_chat",
	"get_contact",
	"get_contact_chats",
	"get_direct_chat_by_contact",
	"get_last_interaction",
	"get_media_notes",
	"get_media_stats",
	"get_message_context",
	"get_notes",
	"get_poll_results",
	"list_chats",
	"list_group_members",
	"list_media",
	"list_messages",
	"list_unanswered",
	"list_unread",
	"message_stats",
	"search_contacts",
	"search_media_notes",
	"search_notes",
	"transcribe_audio",
}

// toolPolicy is the parsed pair of lists. The zero value allows everything.
type toolPolicy struct {
	allow map[string]bool
	deny  map[string]bool
}

// knownTools is every name the two lists accept.
func knownTools() map[string]bool {
	known := make(map[string]bool, len(unenforcedTools)+len(endpointTools))
	for _, name := range unenforcedTools {
		known[name] = true
	}
	for _, tools := range endpointTools {
		for _, name := range tools {
			known[name] = true
		}
	}
	return known
}

func parseToolList(raw string) map[string]bool {
	names := map[string]bool{}
	for _, item := range strings.Split(raw, ",") {
		if item = strings.TrimSpace(item); item != "" {
			names[item] = true
		}
	}
	return names
}

func sortedNames(names map[string]bool) []string {
	out := make([]string, 0, len(names))
	for name := range names {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}

// newToolPolicy parses both lists and rejects names that are not tools.
func newToolPolicy(allowRaw, denyRaw string) (toolPolicy, error) {
	p := toolPolicy{allow: parseToolList(allowRaw), deny: parseToolList(denyRaw)}
	known := knownTools()
	for _, list := range []struct {
		env   string
		names map[string]bool
	}{{allowToolsEnv, p.allow}, {denyToolsEnv, p.deny}} {
		var unknown []string
		for _, name := range sortedNames(list.names) {
			if !known[name] {
				unknown = append(unknown, name)
			}
		}
		if len(unknown) > 0 {
			return toolPolicy{}, fmt.Errorf("%s lists unknown tool(s): %s. Valid names: %s",
				list.env, strings.Join(unknown, ", "), strings.Join(sortedNames(known), ", "))
		}
	}
	return p, nil
}

func loadToolPolicy() (toolPolicy, error) {
	return newToolPolicy(os.Getenv(allowToolsEnv), os.Getenv(denyToolsEnv))
}

// restricted reports whether either list is set (unset = allow everything).
func (p toolPolicy) restricted() bool {
	return len(p.allow) > 0 || len(p.deny) > 0
}

// allowsTool applies deny-then-allow to one tool name.
func (p toolPolicy) allowsTool(name string) bool {
	if p.deny[name] {
		return false
	}
	return len(p.allow) == 0 || p.allow[name]
}

// refusal returns the 403 message for a path, or "" when the path may proceed.
func (p toolPolicy) refusal(path string) string {
	if !p.restricted() {
		return ""
	}
	tools, mapped := endpointTools[path]
	if !mapped {
		// A mutating endpoint that nobody added to endpointTools. Fail closed:
		// an unclassified side effect must not slip past a list that is set.
		return path + " is disabled: it maps to no MCP tool name, so " + allowToolsEnv +
			" / " + denyToolsEnv + " cannot classify it"
	}
	reasons := make([]string, 0, len(tools))
	for _, name := range tools {
		if p.allowsTool(name) {
			return ""
		}
		env := allowToolsEnv + " is set and does not list it"
		if p.deny[name] {
			env = denyToolsEnv + " lists it"
		}
		reasons = append(reasons, name+" ("+env+")")
	}
	return path + " is disabled: every tool that uses it is blocked: " + strings.Join(reasons, "; ")
}

// guard wraps a mutating handler: 403 with the shared JSON error shape when no
// tool that reaches this endpoint is allowed, pass-through otherwise. Applied
// inside readOnlyPolicy.guard, so read-only answers first when both apply.
func (p toolPolicy) guard(h http.HandlerFunc) http.HandlerFunc {
	if !p.restricted() {
		return h
	}
	return func(w http.ResponseWriter, r *http.Request) {
		if msg := p.refusal(r.URL.Path); msg != "" {
			writeError(w, http.StatusForbidden, msg)
			return
		}
		h(w, r)
	}
}

// Summary is a one-line description for the startup log.
func (p toolPolicy) Summary() string {
	if !p.restricted() {
		return allowToolsEnv + " / " + denyToolsEnv + " unset: every endpoint allowed by the other layers stays open"
	}
	parts := []string{}
	if len(p.allow) > 0 {
		parts = append(parts, allowToolsEnv+": "+strings.Join(sortedNames(p.allow), ", "))
	}
	if len(p.deny) > 0 {
		parts = append(parts, denyToolsEnv+": "+strings.Join(sortedNames(p.deny), ", "))
	}
	var blocked []string
	for path := range endpointTools {
		if p.refusal(path) != "" {
			blocked = append(blocked, path)
		}
	}
	sort.Strings(blocked)
	if len(blocked) == 0 {
		return "Endpoint policy — " + strings.Join(parts, "; ") + "; every mutating endpoint still enabled"
	}
	return fmt.Sprintf("Endpoint policy — %s; %d endpoint(s) answer 403 (%s)",
		strings.Join(parts, "; "), len(blocked), strings.Join(blocked, ", "))
}
